### ISSUE TITLE
GraphQL returns 401 (unauthorized) error

### ISSUE THREAD

**Issue by alxnkt** [strapi/strapi#3058]
**Informations**
- **Node.js version**: v10.15.2 
- **NPM version**: 6.9.0
- **Strapi version**: 3.0.0-alpha.24.1 
- **Database**: MongoDB v4.0.6
- **Operating system**: MacOS
- **(Optional) Link to your Project**: 

**What is the current behavior?**
Some time ago I have completed the brilliant tutorial from here https://blog.strapi.io/cooking-a-deliveroo-clone-with-nuxt-vue-js-graphql-strapi-and-stripe-setup-part-1-7/
It worked like a charm!
But when I tried to repeat my success, I suddenly fell into some weird issue. When I first time load a page (server-rendered) everything is going ok. But when I click any link, causing graphql (ajax) request, I get an error:
```json
{
    "errors": [
        {
            "message": "User Not Found",
            "locations": [
                {
                    "line": 2,
                    "column": 13
                }
            ],
            "path": [
                "stories"
            ],
            "extensions": {
                "code": "INTERNAL_SERVER_ERROR",
                "exception": {
                    "data": null,
                    "isBoom": true,
                    "isServer": false,
                    "output": {
                        "statusCode": 401,
                        "payload": {
                            "statusCode": 401,
                            "error": "Unauthorized",
                            "message": "User Not Found"
                        },
                        "headers": {}
                    },
                    "stacktrace": [
                        "Error: User Not Found",
                        "    at handleErrors (/Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/config/policies/permissions.js:73:53)",
                        "    at module.exports (/Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/config/policies/permissions.js:20:14)",
                        "    at process._tickCallback (internal/process/next_tick.js:68:7)"
                    ]
                }
            }
        }
    ],
    "data": {
        "stories": null
    }
}
```
permissions.js:20:14:
```javascript
if (!ctx.state.user) {
  return handleErrors(ctx, 'User Not Found', 'unauthorized');
}
```
permissions.js:73:53:
```javascript
if (ctx.request.graphql === null) {
  return ctx.request.graphql = strapi.errors[type](err);
}
```
As you can see, it tries to get user information, but fails... 
All necessary permissions are granted (including public and authenticated). And i am able to get my data via http://localhost:1337/mycollection 
There is some strange thing, that I have noticed. In my old database (MongoDB) there was a field "id" (not "_id") in collection ```users-permissions_user```:
```
> db.getCollection('users-permissions_user').find();
{ "_id" : ObjectId("5c860504064bb592e0cd6bcd"), 
"confirmed" : true, "blocked" : false, "username" : "admin", "password" : "abrakadabra", 
"email" : "admin@admin.com", "provider" : "local", 
"role" : ObjectId("5c8604ce064bb592e0cd6ae9"), "__v" : 0, 
"id" : "5c860504064bb592e0cd6bcd" }
```
But there is no such a field in my new database:
```
> db.getCollection('users-permissions_user').find();
{ "_id" : ObjectId("5ca0efbd4a5f81df36f8ff9f"), 
"confirmed" : true, "blocked" : false, "username" : "admin", "password" : "abrakadabra", 
"email" : "admin@admin.com", "provider" : "local", 
"role" : ObjectId("5ca0ef8f4a5f81df36f8febb"), "__v" : 0 }
```
When I try to add this field manually, strapi fails to start.
```
[2019-03-31T19:23:39.818Z] debug ⛔️ Server wasn't able to start properly.
[2019-03-31T19:23:39.820Z] error id is not defined
ReferenceError: id is not defined
```
**Steps to reproduce the problem**
Complete tutorial (see above) steps 1-4.

**What is the expected behavior?**
Server (graphql) returns valid JSON response.

**Suggested solutions**
IMHO Return id field in database? Actually, I have no idea ))

**Comment by alxnkt** [comment]
By the way... I am able to get necessary data via curl: 
```
curl -d '{"query":"{graphql_query_text}"}' -H "Content-Type: application/json" -X POST http://localhost:1337/graphql
```
This query returns valid data.

Weird....

**Comment by alxnkt** [comment]
Hooray!
After updating from 3.0.0-alpha.24.1 to 3.0.0-alpha.25.2 I received a new error:
```
{
    "errors": [
        {
            "message": "Invalid token.",
            "locations": [
                {
                    "line": 2,
                    "column": 13
                }
            ],
            "path": [
                "story"
            ],
            "extensions": {
                "code": "INTERNAL_SERVER_ERROR",
                "exception": {
                    "isBoom": true,
                    "isServer": false,
                    "data": null,
                    "output": {
                        "statusCode": 401,
                        "payload": {
                            "statusCode": 401,
                            "error": "Unauthorized",
                            "message": "Invalid token."
                        },
                        "headers": {}
                    },
                    "stacktrace": [
                        "Error: Invalid token.",
                        "    at /Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/services/Jwt.js:58:27",
                        "    at /Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/node_modules/jsonwebtoken/verify.js:133:14",
                        "    at getSecret (/Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/node_modules/jsonwebtoken/verify.js:90:14)",
                        "    at Object.module.exports [as verify] (/Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/node_modules/jsonwebtoken/verify.js:94:10)",
                        "    at /Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/services/Jwt.js:52:11",
                        "    at new Promise (<anonymous>)",
                        "    at Object.verify (/Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/services/Jwt.js:51:12)",
                        "    at Object.getToken (/Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/services/Jwt.js:38:17)",
                        "    at module.exports (/Users/user1/Projects/nodejs/my-project/server/plugins/users-permissions/config/policies/permissions.js:8:82)",
                        "    at dispatch (/usr/local/lib/node_modules/strapi/node_modules/koa-compose/index.js:42:32)",
                        "    at /Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/strapi-utils/lib/policy.js:111:13",
                        "    at dispatch (/usr/local/lib/node_modules/strapi/node_modules/koa-compose/index.js:42:32)",
                        "    at /usr/local/lib/node_modules/strapi/node_modules/koa-compose/index.js:34:12",
                        "    at /Users/user1/Projects/nodejs/my-project/server/plugins/graphql/services/Query.js:238:69",
                        "    at field.resolve (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/graphql-extensions/dist/index.js:128:26)",
                        "    at resolveFieldValueOrError (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/graphql/execution/execute.js:480:18)",
                        "    at resolveField (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/graphql/execution/execute.js:447:16)",
                        "    at executeFields (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/graphql/execution/execute.js:294:18)",
                        "    at executeOperation (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/graphql/execution/execute.js:238:122)",
                        "    at executeImpl (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/graphql/execution/execute.js:85:14)",
                        "    at Object.execute (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/graphql/execution/execute.js:62:35)",
                        "    at /Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/apollo-server-core/dist/requestPipeline.js:198:42",
                        "    at Generator.next (<anonymous>)",
                        "    at /Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/apollo-server-core/dist/requestPipeline.js:7:71",
                        "    at new Promise (<anonymous>)",
                        "    at __awaiter (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/apollo-server-core/dist/requestPipeline.js:3:12)",
                        "    at execute (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/apollo-server-core/dist/requestPipeline.js:182:20)",
                        "    at Object.<anonymous> (/Users/user1/Projects/nodejs/my-project/server/plugins/graphql/node_modules/apollo-server-core/dist/requestPipeline.js:136:35)"
                    ]
                }
            }
        }
    ],
    "data": {
        "story": null
    }
}
```
Looks like I'm on the right way.

**Comment by alxnkt** [comment]
I keep trying!
After some debugging, I found out that strapi itself and nuxt use different JWT tokens. So strapi works pretty good, but nuxt ajax-request fails (nuxt server-side works as well).
```
  verify: (token) => {
    return new Promise(function (resolve, reject) {
      jwt.verify(
        token,
        process.env.JWT_SECRET || _.get(strapi.plugins['users-permissions'], 'config.jwtSecret') || 'oursecret',
        {},
        function (err, tokenPayload = {}) {
          if (err) {
            return reject(new Error('Invalid token.'));
          }
          resolve(tokenPayload);
        }
      );
    });
  }
```
Does anybody know, how to feed proper JWT token to nuxt? Documentation is not clear enough about this theme. I tried to read this: https://strapi.io/documentation/3.x.x/guides/authentication.html
This paper describes, how to register user, get token, and make axios request. But my (only) user already has a token. What do I miss?

**Comment by alxnkt** [comment]
**SOLUTION**
After several days of debugging, googling around and talking to myself, I found out that my question was completely stupid.
The cause of this issue is that nuxt store "client-side" JWT in cookie. So when I accomplished my first tutorial, browser saved old JWT in cookies. When I tried to go through the process one more time, my browser tried to use saved old cookies. But strapi project was new.
**Clearing cookies solved the issue**
Thus, it seems to be more nuxt-related issue, than strapi-related issue. Sorry about that.
Just leave it here.
